components/virtual-node/virtual-node-rpcops-service-impl/src/main/kotlin/net/corda/virtualnode/rpcops/impl/v1/VirtualNodeRestResourceImpl.kt

```
@@ -348,7 +350,7 @@ internal class VirtualNodeRestResourceImpl @Activate constructor(
         if (!isRunning) throw IllegalStateException(
             "${this.javaClass.simpleName} is not running! Its status is: ${lifecycleCoordinator.status}"
         )
-        // TODO: Validate newState
+        validateStateChange(virtualNodeShortId, newState)
```

Performing this check in the REST worker actually isn't enough. We also need to check in the VirtualNodeWriterProcessor when we actually perform the read of the virtual node from the database.

This is because the virtualNodeInfoReadService only provides an eventually consistent version of the data. There is a delay propagating the virtual node info to all of the subscribers from the db-worker. The db-worker has the single source of truth and correct view of the state of the virtual node.

For example, if the state gets changed, rest-worker-A might get the updated vnodeInfo, and block attempts to change the state, but rest-worker-B might not receive the updated vnodeInfo, and will send the request over to the db worker. The db-worker needs to perform this check, as it represents the single source of truth of the true state of the virtual node. REST requests can arrive at any rest worker based on load balancing.

---

I actually think best course of action is to remove the validation from the REST worker, move your validation over to the db-worker to be performed in VirtualNodeWriterProcessor. You can return a new exception in the virtual-node-common and translate that to the InvalidStateChangeException on rest-worker side.

---

https://r3-cev.atlassian.net/browse/CORE-10454 - this validation needs to be in db-worker.
https://r3-cev.atlassian.net/browse/CORE-10459 - this validation can be in rest-worker since it doesn't rely on the virtual node state.
https://r3-cev.atlassian.net/browse/CORE-10403 - this validation can be in rest-worker too

---

grand Ive removed that from the rest resource and Ill move it into the worker. Ill do it in this PR though so hold tight

---

Yeah looks good 👍